cmd/dp-observation-importer/main.go

```diff
@@ -88,18 +105,49 @@ func main() {
 	// handle a batch of events.
 	batchHandler := event.NewBatchHandler(observationMapper, observationStore, resultWriter, errorHandler)
 
+	eventConsumer := event.NewConsumer()
+
 	// Start listening for event messages.
-	go event.Consume(kafkaConsumer, config.BatchSize, batchHandler, config.BatchWaitTime, exit)
+	eventConsumer.Consume(kafkaConsumer, config.BatchSize, batchHandler, config.BatchWaitTime, errorChannel)
+
+	shutdownGracefully := func(err error) {
+
+		if err != nil {
+			log.Error(err, nil)
+		}
 
-	<-signals
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
 
-	close(exit)
+		// gracefully dispose resources
+		eventConsumer.Close(ctx)
+		kafkaConsumer.Close(ctx)
+		kafkaErrorProducer.Close(ctx)
+		kafkaResultProducer.Close(ctx)
+		httpServer.Shutdown(ctx)
 
-	// gracefully dispose resources
-	kafkaConsumer.Closer() <- true
-	kafkaErrorProducer.Closer() <- true
-	kafkaResultProducer.Closer() <- true
+		// cancel the timer in the shutdown context.
+		cancel()
 
-	log.Debug("graceful shutdown was successful", nil)
-	os.Exit(0)
+		log.Debug("graceful shutdown was successful", nil)
+		os.Exit(0)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
+
+	for {
+		select {
+
+		case err := <-kafkaConsumer.Errors():
```

Comment:
I think we should add a log statement for each cases - always useful to know what caused the shutdown. I know we are passing the error into the shutdown func to be logged out but I think it would be useful to see something like "Kafka producer received error - invoking graceful shutdown..."

Reply:
Good point - done